packages/dev/core/src/Misc/snapshotRenderingHelper.ts

```diff
@@ -0,0 +1,234 @@
+            const save = this._engine.snapshotRendering;
+            this._engine.snapshotRendering = false;
+            if (save) {
+                this.enableSnapshotRendering();
+            }
+        });
+
+        this._scene.onBeforeRenderObservable.add(() => {
+            if (!this._engine.snapshotRendering || this._engine.snapshotRenderingMode !== Constants.SNAPSHOTRENDERING_FAST) {
+                return;
+            }
+
+            // Animate skeletons
+            scene.skeletons.forEach((skeleton) => skeleton.prepare(true));
+
+            for (const mesh of scene.meshes) {
+                if (mesh.infiniteDistance) {
+                    mesh.transferToEffect(mesh.computeWorldMatrix(true));
+                }
+
+                if (mesh.skeleton) {
+                    mesh.transferToEffect(mesh.computeWorldMatrix(true));
+                }
+
+                if (mesh.morphTargetManager && mesh.subMeshes) {
+                    // Make sure morph target animations work
+                    for (const subMesh of mesh.subMeshes) {
+                        const dw = subMesh._drawWrapper;
+                        const effect = dw.effect;
+                        if (effect) {
+                            const dataBuffer = (dw.drawContext as WebGPUDrawContext).buffers["LeftOver" satisfies (typeof WebGPUShaderProcessor)["LeftOvertUBOName"]];
+                            const ubLeftOver = (effect._pipelineContext as WebGPUPipelineContext)?.uniformBuffer;
+                            if (dataBuffer && ubLeftOver && ubLeftOver.setDataBuffer(dataBuffer)) {
+                                mesh.morphTargetManager._bind(effect);
+                                BindMorphTargetParameters(mesh, effect);
+                                ubLeftOver.update();
+                            }
+                        }
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Enable snapshot rendering
+     * Use this method instead of engine.snapshotRendering=true, to make sure everything is ready before enabling snapshot rendering.
+     * Note that this method is ref-counted and works in pair with disableSnapshotRendering(): you should call enableSnapshotRendering() as many times as you call disableSnapshotRendering().
+     */
+    public enableSnapshotRendering() {
+        if (!this._engine.isWebGPU) {
+            return;
+        }
+
+        if (--this._disableRenderingRefCount > 0) {
+            return;
+        }
+
+        this._disableRenderingRefCount = 0;
+
+        this._scene.executeWhenReady(() => {
+            if (this._disableRenderingRefCount > 0) {
+                return;
+            }
+
+            // Make sure a full frame is rendered before enabling snapshot rendering, so use "+2" instead of "+1"
+            this._executeAtFrame(this._engine.frameId + 2, () => {
+                this._engine.snapshotRendering = true;
+            });
+        });

```

Comment:
The ref count check needs to be in the inner most part of the asynchrony, right? Otherwise it is still possible for `disableSnapshotRendering` to be called after the ref count check but before snapshotRendering is set to true.
```suggestion
        this._scene.executeWhenReady(() => {
            // Make sure a full frame is rendered before enabling snapshot rendering, so use "+2" instead of "+1"
            this._executeAtFrame(this._engine.frameId + 2, () => {
                if (this._disableRenderingRefCount > 0) {
                      return;
                }

                this._engine.snapshotRendering = true;
            });
        });

```

Reply:
I don't think so, because `_executeAtFrame` does:
```typescript
        const obs = this._engine.onEndFrameObservable.add(() => {
            if (this._disableRenderingRefCount > 0) {
                this._engine.onEndFrameObservable.remove(obs);
                return;
            }
            if (this._engine.frameId >= frameId) {
                this._engine.onEndFrameObservable.remove(obs);
                func();
            }
        });
```
Thus, checking `_disableRenderingRefCount` inside `func` is not necessary, as it is not possible for `_disableRenderingRefCount` to change between the time it is 0 and the time `func` is executed.